### PR TITLE
docs: updates to the seeding documentation

### PR DESCRIPTION
**Pull request by LuisFelipeCoelho**
Some updates to the seeding documentation

### CODE REVIEW

**Comment by tboldagh** [comment]
The document starts with exclaimer that the intro section is outdated. Is it still needed? The intro is fairly general anyways.

File: docs/core/seeding.rst [hunk 1]
```diff
@@ -93,77 +101,103 @@ only once, the circle calculation is spread out over the three loops.
...
-         // calculated radius must not be smaller than minimum radius
-         if (S2 < B2 * m_config.minHelixDiameter2) {
-           continue;
-         }
-         // 1/helixradius: (B/sqrt(S2))/2 (we leave everything squared)
-         float iHelixDiameter2 = B2 / S2;
+	for (auto spM : middleSPs) {
+   
+		... // compatibility cuts between SP duplets
+
```
Comment (paulgessinger) [inline docs/core/seeding.rst:107]: The problems with these code listings is that they have the potential to become outdated very quickly.
Do you see any way you could break this up somehow and describe the sequence in words, while linking to the various functions that are called using the sphinx syntax?

**Comment by paulgessinger** [comment]
> The document starts with exclaimer that the intro section is outdated. Is it still needed? The intro is fairly general anyways.

True, with this PR that should probably be removed.

**Comment by tboldagh** [comment]
Can't comment o unchanged text. But I have simplifying suggestions about captions:
1 - Sketch of the detector with 3 layers. The interaction region is supposed to be located along the z axis and have size significantly smaller than the radius of the innermost detector layer.
2- The x-y projection of the detector with the charged particle helical track originating from the centre of the detector. Signals left by passage of the track through the detector layers are marked with green crosses. 
3- The r-z projection of the detector with the same charged particle track. The track is depicted with the same colours as on previous figure. 

If you have a chance to touch the plots can they be resized so that on each of them the detector has similar size. 
On the figure 3  I would suggest to mark the radius of the detectror as done in 2.

In general the explanations do not belong to the picture captions but should be part of the text.

File: docs/core/seeding.rst [hunk 2]
```diff
@@ -173,7 +207,7 @@ energy and impact parameters can be estimated.
 The scattering calculation is also spread over the nested loops to avoid
 redoing calculations. First, the maximum allowed scattering at the configured
 minimum transverse momentum (pT) cut is calculated and scaled by the
-pseudorapidity of the bottomSP-middleSP dupletto get the minimum momentum of
+pseudorapidity of the bottomSP-middleSP duplet to get the minimum momentum of
```
Comment (tboldagh) [inline docs/core/seeding.rst:210]: SP from bottom layer SP from middle layer

File: docs/core/seeding.rst [hunk 3]
```diff
@@ -194,46 +228,55 @@ scattering for the triplet as well.
...
    scatteringInRegion2 *=
        m_config.sigmaScattering * m_config.sigmaScattering;
+       
+The minimum scattering term ('scatteringInRegion2') is calculated from
+'sigmaScattering', the configurable number of sigmas of scattering angle
+to be considered, and 'maxScatteringAngle2', which is evaluated from the
+Lynch & Dahl correction of the Highland equation assuming the lowest
+allowed pT. The parameters of the Highland equation are fully configurable.
 
 The following code block calculates if the triplet forms a nearly straight line
```
Comment (tboldagh) [inline docs/core/seeding.rst:238]: calculates -> checks

File: docs/core/seeding.rst [hunk 4]
```diff
@@ -194,46 +228,55 @@ scattering for the triplet as well.
...
 in the r/z plane (see :numref:`rz`) as the particle path in the r/z plane is
 unaffected by the magnetic field [#f1]_. This is split in two (may be revised
```
Comment (tboldagh) [inline docs/core/seeding.rst:240]: tow parts

File: docs/core/seeding.rst [hunk 5]
```diff
@@ -194,46 +228,55 @@ scattering for the triplet as well.
...
+		if (m_config.skipPreviousTopSP) {
+			// break if cotTheta from bottom SP < cotTheta from top SP because
+			// the SP are sorted by cotTheta
+			if (cotThetaB - cotThetaT < 0) {
+				break;
+			}
+			t0 = t + 1;
+		}
+		continue;
+	}
```
Comment (tboldagh) [inline docs/core/seeding.rst:279]: As suggested above. Better not to have the code here. Gets out of date quickly. Pseudocode in some form is better.

File: docs/core/seeding.rst [hunk 6]
```diff
@@ -194,46 +228,55 @@ scattering for the triplet as well.
...
 
 Now the check for scattering with calculated particle momentum. Momentum is
```
Comment (tboldagh) [inline docs/core/seeding.rst:281]: Next/Following check takes into account estimate particle momentum (smaller scattering angle is permitted for higher momentum) and pseudorapidity (larger scattering takes into account amount of the material crosses that takes depends on the angle). (If that is the reason actually.)

File: docs/core/seeding.rst [hunk 7]
```diff
@@ -242,19 +285,34 @@ previous check, as the scattering decreases linearly with particle energy
...
+		if (m_config.skipPreviousTopSP) {
+			if (cotThetaB - cotThetaT < 0) {
+				break;
+			}
+			t0 = t;
+		}
+		continue;
+	}
 
 The last cut applied in this function is on the so-called impact parameters,
```
Comment (tboldagh) [inline docs/core/seeding.rst:317]: transverse impact parameter (or DCA - distance of closest approach) - various experiments refer to it differently

File: docs/core/seeding.rst [hunk 8]
```diff
@@ -263,18 +321,30 @@ compatible with both the current middle SP and current bottom SP.
...
+
+		// evaluate eta and pT of the seed
+		float cotThetaAvg = std::sqrt(cotThetaAvg2);
+		float theta = std::atan(1. / cotThetaAvg);
+		float eta = -std::log(std::tan(0.5 * theta));
+		state.etaVec.push_back(eta);
+		state.ptVec.push_back(pT);
+	}
 
 The bottom SP and middle SP as well as the collection of top SP is passed to
```
Comment (tboldagh) [inline docs/core/seeding.rst:349]: Sorry, can't understand this. Maybe simply start with saying that the seeds are filtered after being crated. And instead of details of function calls explain why and how?

File: docs/core/seeding.rst [hunk 9]
```diff
@@ -288,19 +358,20 @@ This function assigns a weight (which should correspond to the likelihood that
 a seed is good) to all seeds and calls the detector specific cuts to apply a
```
Comment (tboldagh) [inline docs/core/seeding.rst:358]: calls -> applies detector specific section of seeds based on weights. (not clear what the weight modifications is for)

File: docs/core/seeding.rst [hunk 10]
```diff
@@ -288,19 +358,20 @@ This function assigns a weight (which should correspond to the likelihood that
...
-SeedFilter::filterSeeds_1SpFixed
+SeedFilter::filterSeeds_1SpFixed. This process is important to improving computational
+performance and the quality of the final track collections by rejecting lower-quality seeds.
 
-The weight is influenced by:
+The weight can be influenced by:
 
-#. The impact parameter (the higher the distance the worse)
-#. The number of seeds which may belong to the same particle track
+#. The transverse (`d_{0}`) and longitudinal (`z_{0}`) impact parameters (the higher the distance the worse)
```
Comment (tboldagh) [inline docs/core/seeding.rst:366]: the worse -> the smaller the weight

File: docs/core/seeding.rst [hunk 11]
```diff
@@ -288,19 +358,20 @@ This function assigns a weight (which should correspond to the likelihood that
...
+#. The number of seeds which may belong to the same particle track (`N_{t}`)
```
Comment (tboldagh) [inline docs/core/seeding.rst:367]: how is the weight impacted by this?

File: docs/core/seeding.rst [hunk 12]
```diff
@@ -288,19 +358,20 @@ This function assigns a weight (which should correspond to the likelihood that
...
-The impact parameter is multiplied by the configured factor and subtracted from
+The transverse impact parameter is multiplied by the configured factor and subtracted from
 the weight, as seeds with higher impact parameters are assumed to be less
 likely to stem from a particle than another seed using the same middle SP with
-smaller impact parameters.
+smaller impact parameters. The longitudinal impact parameter is subtracted from
+the weight if configured.
 
 The number of seeds only differing in top SP which have similar helix radius
 and the same sign (i.e. the same charge) is used to increase the weight, as it
```
Comment (tboldagh) [inline docs/core/seeding.rst:377]: text below introduces "measurement". I suggest to stay with SP. In any case does this cover SPs in the overlapping modules? The logic below is quite hard to follow.

File: docs/core/seeding.rst [hunk 13]
```diff
@@ -313,6 +384,36 @@ configured factor and added to the weight.
 The optional detector specific cuts can use the weight from 1. and 2. and the
```
Comment (tboldagh) [inline docs/core/seeding.rst:384]: Until now it seemed that there is only one weight?

File: docs/core/seeding.rst [hunk 14]
```diff
@@ -313,6 +384,36 @@ configured factor and added to the weight.
...
 three SP to apply a hard cut or change the weight of a seed.
 
+This function also includes a fully configurable seed confirmation step that, when enabled
```
Comment (tboldagh) [inline docs/core/seeding.rst:387]: Is this in addition to above? Need to explain the relation of these filters.

File: docs/core/seeding.rst [hunk 15]
```diff
@@ -324,6 +425,12 @@ location. The ordering by location is only done to make sure reimplementations
 (such as the GPU code) are comparable and return the bitwise exactly same
 result.
 
+When a seed is accepted and seed confirmation is enabled, the weight of that seed
+is assigned to each of its SPs. Each SP will hold the weight of the best seed that
+includes that SP. This information is used in the selection of the next seeds:
```
Comment (tboldagh) [inline docs/core/seeding.rst:430]: next seeds?

File: docs/core/seeding.rst [hunk 16]
```diff
@@ -324,6 +425,12 @@ location. The ordering by location is only done to make sure reimplementations
...
+The seed is kept only if its weight is greater than the weight of at least one of
```
Comment (tboldagh) [inline docs/core/seeding.rst:431]: Maybe start by saying that it is some sort of voting scheme that is implemented here. BTW. Greater or equal, right? 
I.e. in the illustrations above there will be only one weight value.

**Review by tboldagh — CHANGES_REQUESTED** [review]

**Comment by LuisFelipeCoelho** [comment]
Thanks a lot @tboldagh for your comments! I will update this ASAP

File: docs/core/seeding.rst [hunk 17]
```diff
@@ -11,7 +11,7 @@ find all measurements belonging to a single particle in this direction in order
 to reconstruct the track. This means, if no seed exists for a particle, this
 particle will not be reconstructed. On the other hand, finding too many seeds
 which either correspond to a particle for which another seed already exists or
-which does not correspond to a particle at all increases the time needed for
+which do not correspond to a particle at all increases the time needed for
```
[the remark this answers is not in the document]
Comment (LuisFelipeCoelho) [reply docs/core/seeding.rst]: I thought this referred to "too many seeds" from "finding too many seeds".

**Comment by LuisFelipeCoelho** [comment]
@tboldagh @paulgessinger 
I hope that I have addressed all the comments.
I have decided to replace the code block with some equations and further explanation. If you think that I have added too much information or that the pseudo code is still necessary or that I need to add more information, please tell me and I will change it accordingly.

File: docs/core/seeding.md [hunk 18]
```diff
@@ -0,0 +1,234 @@
+# Track Seeding
```
Comment (paulgessinger) [inline docs/core/seeding.md:1]: The rST had an anchor that was linked to:

```suggestion
(seeding_core)=

# Track Seeding
```

File: docs/core/seeding.md [hunk 19]
```diff
@@ -0,0 +1,234 @@
...
+(tripletsFormation)=
+:::{figure} ../figures/seeding/tripletsFormation.svg
+:width: 450px
+:align: center
+Representation of the search for triplet combinations in the $(r, z)$ plane. The bins used in the search are represented in different colours.
+:::
+
+### The Seed Finder
+
+The The {class}`SeedFilter::createSeedsForGroup` function receives three iterators
```
Comment (paulgessinger) [inline docs/core/seeding.md:58]:
```suggestion
The {func}`SeedFilter::createSeedsForGroup` function receives three iterators 
```

File: docs/core/seeding.md [hunk 20]
```diff
@@ -0,0 +1,234 @@
...
+
+\begin{equation*}
+d_0 \leq \left| \left( A - B \cdot r_M \right) \cdot r_M \right|
+\end{equation*}
+
+### The Seed Filter
+
+After creating the potential seeds we apply a seed filter procedure that compares the seeds with other SPs compatible with the seed curvature.
+This process ranks the potential seeds based on certain quality criteria and selects the ones that are more likely to produce high-quality tracks 
+The filter is divided into two functions {class}`SeedFilter::filterSeeds_2SpFixed` and {class}`SeedFilter::filterSeeds_1SpFixed`.
```
Comment (paulgessinger) [inline docs/core/seeding.md:181]:
```suggestion
The filter is divided into two functions {func}`SeedFilter::filterSeeds_2SpFixed` and {func}`SeedFilter::filterSeeds_1SpFixed`.
```

File: docs/core/seeding.md [hunk 21]
```diff
@@ -0,0 +1,234 @@
...
+the weight, as seeds with higher impact parameters are assumed to be less
+likely to stem from a particle than another seed using the same middle SP with
+smaller impact parameters. The number of compatible seeds ($N_t$) is used to increase the weight, as a higher number of measurements 
+will lead to higher quality tracks.  Finally, the weight can also be affected by optional detector-specific cuts.
+
+:::{note} 
+The $z_0$ term in the weight is not yet fully configurable, but this will change soon.
+:::
+
+The {class}`SeedFilter::filterSeeds_2SpFixed` function also includes a configurable `Acts::SeedConfirmationRangeConfig` seed confirmation step that, when enabled,
```
Comment (paulgessinger) [inline docs/core/seeding.md:212]:
```suggestion
The {func}`SeedFilter::filterSeeds_2SpFixed` function also includes a configurable {class}`Acts::SeedConfirmationRangeConfig` seed confirmation step that, when enabled,
```

File: docs/core/seeding.md [hunk 22]
```diff
@@ -0,0 +1,234 @@
...
+After creating the potential seeds we apply a seed filter procedure that compares the seeds with other SPs compatible with the seed curvature.
+This process ranks the potential seeds based on certain quality criteria and selects the ones that are more likely to produce high-quality tracks 
+The filter is divided into two functions {class}`SeedFilter::filterSeeds_2SpFixed` and {class}`SeedFilter::filterSeeds_1SpFixed`.
+
+The first function compares the middle and bottom layer SPs of the seeds to other top layer SPs; seeds only differing in top SP are 
+compatible if they have similar helix radius with the same sign (i.e. the same charge). The SPs must have a minimum distance in 
+detector radius, such that SPs from the same layer cannot be considered compatible. The second function iterates over the seeds with 
+only a common middle layer SP and selects the higher quality combinations.
+
+#### SeedFilter::filterSeeds_2SpFixed
```
[the remark this answers is not in the document]
Comment (LuisFelipeCoelho) [reply docs/core/seeding.md:190]: Thank you for the comments

**Comment by LuisFelipeCoelho** [comment]
not sure why the docs are failing

**Comment by paulgessinger** [comment]
I think its

```
/__w/acts/acts/docs/core/seeding.md:190: WARNING: doxygenfunction: Cannot find function "SeedFilter::filterSeeds_2SpFixed" in doxygen xml output for project "Acts" from directory: _build/doxygen-xml
/__w/acts/acts/docs/core/seeding.md:226: WARNING: doxygenfunction: Cannot find function "SeedFilter::filterSeeds_1SpFixed" in doxygen xml output for project "Acts" from directory: _build/doxygen-xml
```

I'm guessing you need to add the `Acts::` namespace before the functions?

File: docs/core/seeding.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (AJPfleger) [reply docs/core/seeding.rst]: `which either correspond to particles with already existing seeds or
which do not correspond to particles at all increases the time needed for`

I think it should be plural. Maybe rewording the first sentence could make it clearer.
Also, I think `particles` should be plural.

File: docs/core/seeding.md [hunk 23]
```diff
@@ -0,0 +1,240 @@
...
+
+## Acts Implementation
+
+The seeding implementation in `Core/include/Acts/Seeding/` was written with a focus on parallelism and maintainability and as detector agnostic as possible, only assuming a (near) homogeneous magnetic field with particles originating from the central detector region. Cuts are configurable and can be plugged in as algorithm which is called by the seeding. The seeding works on measurements or “SpacePoints” (SP), which need to provide $(x,y,z)$ coordinates with the $z$ axis being along the magnetic field, and $x$ and $y$. For the seeding algorithm to function the particle point of origin has to have a radius smaller than the radius of the detector layer closest to the interaction region. In other words, this seeding algorithm is not suitable for secondary particles originating far from the interaction region.
```
Comment (AJPfleger) [inline docs/core/seeding.md:25]:
```suggestion
The seeding implementation in `Core/include/Acts/Seeding/` was written with a focus on parallelism and maintainability and as detector agnostic as possible, only assuming a (near) homogeneous magnetic field with particles originating from the central detector region. Cuts are configurable and can be plugged in as an algorithm which is called by the seeding. The seeding works on measurements or “SpacePoints” (SP), which need to provide $(x,y,z)$ coordinates with the $z$ axis being along the magnetic field, and $x$ and $y$. For the seeding algorithm to function the particle point of origin has to have a radius smaller than the radius of the detector layer closest to the interaction region. In other words, this seeding algorithm is not suitable for secondary particles originating far from the interaction region.
```

File: docs/core/seeding.md [hunk 24]
```diff
@@ -0,0 +1,240 @@
...
+bottom SP. 
+
+(impactParameter)=
+:::{figure} ../figures/seeding/impactParameter.svg
+:width: 400px
+:align: center
+oi
+:::
+
+Assuming the middle layer SP in the origin of the $x/y$ frame, as in {numref}`impactParameter`.
```
Comment (AJPfleger) [inline docs/core/seeding.md:166]:
```suggestion
Assuming the middle layer SP is in the origin of the $x/y$ frame, as in {numref}`impactParameter`. 
```

File: docs/core/seeding.md [hunk 25]
```diff
@@ -0,0 +1,240 @@
...
+
+(impactParameter)=
+:::{figure} ../figures/seeding/impactParameter.svg
+:width: 400px
+:align: center
+oi
+:::
+
+Assuming the middle layer SP in the origin of the $x/y$ frame, as in {numref}`impactParameter`. 
+The distance between the center of the helix and the interaction point (IP) is given by
```
Comment (AJPfleger) [inline docs/core/seeding.md:167]:
```suggestion
The distance between the centre of the helix and the interaction point (IP) is given by
```
Since we are using Br.Eng. and to match l.78

**Comment by LuisFelipeCoelho** [comment]
Thanks a lot @AJPfleger for your comments

File: docs/core/seeding.rst (hunk 17 above)
[the remark this answers is not in the document]
Comment (LuisFelipeCoelho) [reply docs/core/seeding.rst]: done

**Comment by LuisFelipeCoelho** [comment]
Thanks for your revision @benjaminhuth
